compiler/rustc_parse/src/parser/diagnostics.rs

```diff
@@ -640,6 +640,34 @@ impl<'a> Parser<'a> {
             }
         }
 
+        // Extra info for `c"str"` before 2021 edition or `c "str"` in all editions. The heuristic
+        // here is to identify a cooked, uninterpolated `c` id immediately followed by a string, or
+        // a cooked, uninterpolated `cr` id immediately followed by a string or a `#`, in an
+        // edition where c-string literals are not allowed. There is the slight possibility of a
+        // false positive for a `cr#` that wasn't intended to start a c-string literal, but the
+        // lexer was greedy and didn't preserve whether the `r#` on its own would have started a
+        // valid raw string literal.
+        if ((self.prev_token.kind == TokenKind::Ident(sym::c, false)
+            && matches!(&self.token.kind, TokenKind::Literal(token::Lit { kind: token::Str, .. })))
+            || (self.prev_token.kind == TokenKind::Ident(sym::cr, false)
+                && matches!(
+                    &self.token.kind,
+                    TokenKind::Literal(token::Lit { kind: token::Str, .. }) | token::Pound
+                )))
+            && self.prev_token.span.hi() == self.token.span.lo()
+            && !self.token.span.at_least_rust_2021()
+        {
+            err.cancel();
```

Comment:
Actually, since this subdiagnostic can have false positives, it might be wiser to keep the “token soup” error, not sure. When I suggested canceling the error, I assumed the diagnostic would be 100% accurate and we could even replace the main message of the diagnostic similar to `error[E0670]: `async fn` is not permitted in Rust 2015`.

Your call though. If you'd like to keep the shorter message, I'd go with `unexpected token` (dropping the `found`) and with `unexpected token` over `found here` since the current phrasing isn't used in any other of rustc's error messages or only very rarely regarding the latter.